src/topology/uniform_space/uniform_convergence.lean

```diff
@@ -0,0 +1,334 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import topology.uniform_space.basic
+
+/-!
+# Uniform convergence
+
+A sequence of functions `Fₙ` (with values in a metric space) converges uniformly on a set `s` to a
+function `f` if, for all `ε > 0`, for all large enough `n`, one has for all `y ∈ s` the inequality
+`dist (f y, Fₙ y) < ε`. Under uniform convergence, many properties of the `Fₙ` pass to the limit,
+most notably continuity. We prove this in the file, defining the notion of uniform convergence
+in the more general setting of uniform spaces, and with respect to an arbitrary indexing set
+endowed with a filter (instead of just `ℕ` with `at_top`).
+
+## Main results
+
+Let `α` be a topological space, `β` a uniform space, `Fₙ` and `f` be functions from `α`to `β`
+(where the index `n` belongs to an indexing type `ι` endowed with a filter `p`).
+
+* `tendsto_uniformly_on F f p s`: the fact that `Fₙ` converges uniformly to `f` on `s`. This means
+  that, for any entourage `u` of the diagonal, for large enough `n` (with respect to `p`), one has
+  `(f y, Fₙ y) ∈ u` for all `y ∈ s`.
+* `tendsto_uniformly F f p`: same notion with `s = univ`.
+* `tendsto_uniformly_on.continuous_on`: a uniform limit on a set of functions which are continuous
+  on this set is itself continuous on this set.
+* `tendsto_uniformly.continuous`: a uniform limit of continuous functions is continuous.
+* `tendsto_uniformly_on.tendsto_comp`: If `Fₙ` tends uniformly to `f` on a set `s`, and `gₙ` tends
+  to `x` within `s`, then `Fₙ gₙ` tends to `f x` if `f` is continuous at `x` within `s`.
+* `tendsto_uniformly.tendsto_comp`: If `Fₙ` tends uniformly to `f`, and `gₙ` tends to `x`, then
+  `Fₙ gₙ` tends to `f x`.
+
+We also define notions where the convergence is locally uniform, called
+`tendsto_locally_uniformly_on F f p s` and `tendsto_locally_uniformly F f p`. The previous theorems
+all have corresponding versions under locally uniform convergence.
+
+## Implementation notes
+
+Most results hold under weaker assumptions of locally uniform approximation. In a first section,
+we prove the results under these weaker assumptions. Then, we derive the results on uniform
+convergence from them.
+
+## Tags
+
+Uniform limit, uniform convergence, tends uniformly to
+ -/
+
+noncomputable theory
+open_locale topological_space classical uniformity
+
+open set filter
+
+universes u v w
+variables {α : Type u} {β : Type v} {γ : Type w}
+
+/-!
+### Different notions of uniform convergence
+
+We define uniform convergence and locally uniform convergence, on a set or in the whole space.
+-/
+
+variables {ι : Type*} [uniform_space β]
+{F : ι → α → β} {f : α → β} {s s' : set α} {x : α} {p : filter ι} {g : ι → α}
+
+/-- A sequence of functions `Fₙ` converges uniformly on a set `s` to a limiting function `f` if,
+for any entourage of the diagonal `u`, one has eventually `(f x, Fₙ x) ∈ u` for all `x ∈ s`. -/
```

Comment:
Shouldn't this docstring and the next one mention `p`? I know this is a lot of data...

Reply:
Since these are important definitions, I agree that the docstrings should be crystal-clear. I have added the data about `p` here and below.